### PR TITLE
feat: use the dominant color of a user's avatar as the color of his/her racing bar

### LINKED ISSUE #723: [Feature] extract info from users' avatars to color the racing bars
### Description

We just released v1.9.0 with a new feature "Contributors Activity Racing Bars". It is a great feature but the random colors it presents are actually a little ugly.

A good idea is to extract color info from users' avatars.

Notes: Bots have no avatars.

### PR DESCRIPTION
**Pull request by tyn1998**
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- resolve #723 

## Details

1. Updated the Node.js version requirement to be >= 18 in the `.nvmrc` file.
2. Use the dominant color of a user's avatar as the color of their racing bar (with the help of [colorthief](https://github.com/lokesh/color-thief?tab=readme-ov-file)). This enhances the visual representation of the racing bar.
3. Since avatar loading and color extracting cost much time, a preload & cache mechanism is introduced.
4. Introduce antd and use its [Spin](https://ant-design.antgroup.com/components/spin) component to present the loading animation.

https://github.com/hypertrons/hypertrons-crx/assets/32434520/eb1c1e1f-b86d-4a5b-90ae-dc65eaefad94

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

### CODE REVIEW

**Comment by tyn1998** [comment]
Loading images and extracting colors need time, which causes the animation to freeze for a while. I think all avatars should be loaded and processed before the animation starts.

File: src/pages/ContentScripts/features/repo-activity-racing-bar/AvatarColorStore.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  public async getColors(loginId: LoginId): Promise<Color[]> {
+    if (!this.colorCache.has(loginId)) {
+      let colors: Color[];
+      // a single white color causes error: https://github.com/lokesh/color-thief/issues/40#issuecomment-802424484
+      try {
+        colors = await this.loadAvatar(loginId)
+          .then((img) =>
+            this.colorThief.getPalette(img, COLOR_COUNT, COLOR_QUALITY)
+          )
```
Comment (TieWay59) [inline src/pages/ContentScripts/features/repo-activity-racing-bar/AvatarColorStore.ts:41]: 虽然这里采用了缓存，但这里实际的性能瓶颈应该在于单线程等待多个图片下载+parse，当作同步任务这样造成了卡顿。比较合理的做法应该是立即在缓存里设置一个默认颜色，再注册异步事件去下载+parse在回调时永久性更新的缓存颜色。保证竞赛图更新的时候立即获取到默认值，假如下载顺利-下次update的时候能够获取到真实的颜色。但是在当前的框架下，看起来改动有所难度。

**Comment by TieWay59** [comment]
> Loading images and extracting colors need time, which causes the animation to freeze for a while. I think all avatars should be loaded and processed before the animation starts.

预处理确实可以绕过现有问题，但假如预处理如果还是同步地处理，只是把延迟时间转移到了动画启动之前。压测一个1000人的仓库估计够呛，（且不提缓存收缩是否值得考虑）。

Preprocessing may solve the current issue, but if it is done synchronously, it only shifts the delay to before the animation begins. I doubt your system can handle a repo of 1000 people with this approach.

File: src/pages/ContentScripts/features/repo-activity-racing-bar/AvatarColorStore.ts (hunk 1 above)
Comment (TieWay59) [reply src/pages/ContentScripts/features/repo-activity-racing-bar/AvatarColorStore.ts:41]: 我甚至觉得更好的该是：AvatarColorStore 在回调中直接修改 bar 的颜色。updateMonth 只在首次发现没颜色的时候向AvatarColorStore 发起颜色请求。

**Review by TieWay59 — CHANGES_REQUESTED** [review]

**Comment by tyn1998** [comment]
@TieWay59 thanks for reviewing the code ❤️ . I have improved the preload and cache mechanism, as well as the UI experience :)

This PR is ready, request for your review ❤️  @lhbvvvvv @wxharry

**Comment by tyn1998** [comment]
Hello guys @zhaosj975 @Tenth-crew @wangyantong2000, would you like to `gh pr checkout 727` (install `gh` first) this PR branch and have a test in your machine?

Reference: https://github.com/hypertrons/hypertrons-crx#contributing

**Comment by Tenth-crew** [comment]
好的师兄，之后会尝试

> Hello guys @zhaosj975 @Tenth-crew @wangyantong2000, would you like to `gh pr checkout 727` (install `gh` first) this PR branch and have a test in your machines? Welcome to your snapshots or recordings.
> 
> Reference: https://github.com/hypertrons/hypertrons-crx#contributing

**Comment by tyn1998** [comment]
@Tenth-crew  Great!

BTW, in this project we use English to communicate with each other so more people abroad can understand and participate in :)

**Comment by zhaosj975** [comment]
image
so convenient it is!

**Comment by Tenth-crew** [comment]
image

**Comment by wangyantong2000** [comment]
Uploading test.mp4…

**Review by lhbvvvvv — APPROVED** [review]
LGTM! 

/approve

**Pull request merged**